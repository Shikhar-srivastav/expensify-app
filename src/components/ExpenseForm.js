import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        }
    }
    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = e => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = createdAt => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = e => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please enter valid details!' }));
        } else {
            this.setState(() => ({ error: undefined }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit} className='form'>
                    {!!this.state.error && <p>{this.state.error}</p>}
                    <div className='input-group__item'>
                        <input
                            className='text-input'
                            type='text'
                            placeholder='Description'
                            autoFocus
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                        />
                    </div>
                    <div className='input-group__item'>
                        <input
                            className='text-input'
                            type='text'
                            placeholder='Amount'
                            value={this.state.amount}
                            onChange={this.onAmountChange}
                        />
                    </div>
                    <div className='input-group__item'>
                        <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={day => false}
                        />
                    </div>
                    <div className='input-group__item'>
                        <textarea
                            className='text-area'
                            placeholder='Add a note for your expense (optional)'
                            value={this.state.note}
                            onChange={this.onNoteChange}
                        ></textarea>
                    </div>
                    <div>
                        <button className='button'>
                            Save Expense
                        </button>
                    </div>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;