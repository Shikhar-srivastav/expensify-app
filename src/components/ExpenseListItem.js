import React from 'react';
import { Link, resolvePath } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = props => (
    <Link to={`/edit/${props._id}`}>
        <div className='list-item'>
            <div>
                <h3>{props.description}</h3>
                <span className='list-item__date'>
                    {moment(props.createdAt).format('Do MMMM YYYY')}
                </span>
            </div>
            <div>
                <h3 className='list-item__price'>Rs{numeral(props.amount).format('0,0.00')}</h3>
            </div>
        </div>
    </Link>
);

export default ExpenseListItem;