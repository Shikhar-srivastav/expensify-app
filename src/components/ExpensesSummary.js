import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getFilteredExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/total';

export const ExpensesSummary = props => (
    <div className='page-header'>
        <div className='container'>
            <p>
                Viewing <span>{props.total.count}</span> expense(s) totalling <span>Rs{numeral(props.total.amount).format('0,0.00')}</span>
            </p>
            <div>
                <Link to='/create'>
                    <button className='button'>Add Expense</button>
                </Link>
            </div>
        </div>
     </div>
);

const mapStatetoProps = (state) => ({
    total: getTotalExpenses(getFilteredExpenses(state.expenses, state.filters)),
});

export default connect(mapStatetoProps)(ExpensesSummary)