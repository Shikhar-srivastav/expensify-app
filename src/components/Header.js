import React from "react";
import { NavLink } from "react-router-dom";

// <NavLink to="help">Help</NavLink>

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/">Dashboard</NavLink>
        <NavLink to="create">Add Expense</NavLink>
    </header>
);

export default Header;