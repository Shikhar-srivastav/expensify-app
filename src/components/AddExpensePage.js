import React from "react";
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

export class AddExpensePage extends React.Component {
    onSubmit = expense => {
        this.props.startAddExpense(expense);
        this.props.navigate('/dashboard');
    }
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='container'>
                        <h1 className='page-title'>Add Expense</h1>
                    </div>
                </div>
                <div className='container'>
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => ({
    navigate: useNavigate(),
    startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);