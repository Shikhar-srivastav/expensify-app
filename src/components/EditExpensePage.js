import React from 'react';
import { connect } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';
import ExpenseForm from '../components/ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = expense => {
        this.props.editExpense(this.props.params._id, expense);
        this.props.navigate('/dashboard');
    }
    onClick = () => {
        this.props.removeExpense(this.props.params._id);
        this.props.navigate('/dashboard');
    }
    currentExpense = this.props.expenses.find(({ _id }) => (
        _id === this.props.params._id
    ));
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='container'>
                        <h1 className='page-title'>Edit Expense</h1>
                    </div>
                </div>
                <div className='container'>
                    <ExpenseForm
                        expense={this.currentExpense}
                        onSubmit={this.onSubmit}
                    />
                    <button
                        className='button__secondary'
                        onClick={this.onClick}
                    >Remove Expense</button>
                </div>
            </div>
        );
    }
}



const mapDispatchToProps = (dispatch) => ({
    editExpense: (_id, expense) => {
        dispatch(startEditExpense(_id, expense));
    },
    removeExpense: (_id) => {
        dispatch(startRemoveExpense(_id));
    },
    navigate: useNavigate(),
    params: useParams()
});

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);