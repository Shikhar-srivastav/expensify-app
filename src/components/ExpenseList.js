import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getFilteredExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className='container'>
        <div className='list-header'>
            <div className='show-for-mobile'>Expenses</div>
            <div className='show-for-desktop'>Expense</div>
            <div className='show-for-desktop'>Amount</div>
        </div>
        {
            props.expenses.length === 0 ? (
                <p className='message'>NO EXPENSES</p>
            ) : (
                props.expenses.map(expense => (
                    <ExpenseListItem
                        key={expense._id}
                        {...expense}
                    />
                ))
            )
        }
    </div>
);

const ConnectedExpenseList = connect(state => {
    return {
        expenses: getFilteredExpenses(state.expenses, state.filters)
    };
})(ExpenseList);

export default ConnectedExpenseList;