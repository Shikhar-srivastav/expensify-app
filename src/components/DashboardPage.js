import React from "react";
import ExpenseList from './ExpenseList';
import ExpenseFilters from './ExpenseFilters';
import ExpensesSummary from './ExpensesSummary';

const DashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;