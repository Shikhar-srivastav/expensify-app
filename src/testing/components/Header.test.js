import React from 'react';
// import ReactShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('Render header for the app', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
});

test('Call start logout on button click', () => {
    const logoutSpy = jest.fn();
    const wrapper = shallow(<Header startLogout={logoutSpy} />);
    wrapper.find('button').simulate('click');
    expect(logoutSpy).toHaveBeenCalled();
});