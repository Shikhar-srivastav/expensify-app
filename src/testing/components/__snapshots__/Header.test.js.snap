// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render header for the app 1`] = `
<header>
  <div
    className="container"
  >
    <div
      className="header__container"
    >
      <Link
        className="header__title"
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="header__button"
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
